pkg/model/multiversion/mapper.go

```diff
@@ -0,0 +1,194 @@
+	ErrAPIVersionNotFound   = errors.New("api version not found")
+	ErrAPIVersionDeprecated = errors.New("api version deprecated")
+)
+
+// ModelsMapper is a multi-version model mapper. It is containing the mapping
+// of each non-deprecated version with their correspending ackmodel.model
+// and APIInfos.
+type ModelsMapper struct {
+	gitRepo *git.Repository
+
+	hubVersion    string
+	spokeVersions []string
+
+	deprecatedVersions []string
+	removedVersions    []string
+
+	apiInfos      map[string]APIInfo
+	modelsMapping map[string]*ackmodel.Model
+}
+
+// NewModelsMapper returns a new Model struct.
+func NewModelsMapper(
+	sdkCacheDir string,
+	serviceAlias string,
+	hubVersion string,
+	apisInfo map[string]APIInfo,
+	defaultConfig ackgenconfig.Config,
+) (*ModelsMapper, error) {
+	if len(apisInfo) == 0 {
+		return nil, fmt.Errorf("empty apisInfo")
+	}
+
+	spokeVersions := make([]string, 0, len(apisInfo)-1)
+	gitRepo, err := util.LoadRepository(sdkCacheDir)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read sdk git repository: %v", err)
+	}
+
+	SDKAPIHelper := ackmodel.NewSDKHelper(sdkCacheDir)
+
+	// create model for each non-deprecated api version
+	modelsMapping := make(map[string]*ackmodel.Model, len(apisInfo))
+	for apiVersion, apiInfo := range apisInfo {
+		// TODO(a-hilaly) filter deprecated and removed api versions and store them
+		// in `multiversion.Model`
+		if apiVersion != hubVersion {
+			spokeVersions = append(spokeVersions, apiVersion)
+		}
+
+		err = SDKAPIHelper.WithSDKVersion(apiInfo.AWSSDKVersion)
+		if err != nil {
+			return nil, err
+		}
+
+		SDKAPI, err := SDKAPIHelper.API(serviceAlias)
+		if err != nil {
+			return nil, err
+		}
+
+		i, err := ackmodel.New(
+			SDKAPI,
+			apiVersion,
+			apiInfo.GeneratorConfigPath,
+			defaultConfig,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create model for apiVersion %s: %v", apiVersion, err)
+		}
+		modelsMapping[apiVersion] = i
+	}
+
+	sort.Strings(spokeVersions)
+	model := &ModelsMapper{
+		gitRepo:       gitRepo,
+		hubVersion:    hubVersion,
+		spokeVersions: spokeVersions,
+		apiInfos:      apisInfo,
+		modelsMapping: modelsMapping,
+	}
+	// TODO(hilalymh): Audit deprecated and removed versions
+	return model, nil
+}
+
+// GetModel returns the model of a given api version.
+func (m *ModelsMapper) GetModel(apiVersion string) (*ackmodel.Model, error) {
+	if err := m.VerifyAPIVersions(apiVersion); err != nil {
+		return nil, fmt.Errorf("cannot verify apiVersions %s: %v", apiVersion, err)
+	}
+	return m.modelsMapping[apiVersion], nil
+}
+
+// GetSpokeVersions returns the spokes versions list.
+func (m *ModelsMapper) GetSpokeVersions() []string {
+	return m.spokeVersions
+}
+
+// GetHubVersion returns the hub version.
+func (m *ModelsMapper) GetHubVersion() string {
+	return m.hubVersion
+}
+
+// CompareHubWith compares a given api version with the hub version and returns
+// slices of FieldDeltas representing the diff between CRDs status and spec fields.
```

Comment:
docstring says the function returns slices of FieldDeltas, but it seems to return a map of string to CRDDelta.